cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/AdapterService.java

```diff
@@ -301,12 +302,24 @@ public void removeAdapter(Id.Namespace namespace, String adapterName)
     throws NotFoundException, CannotBeDeletedException {
 
     AdapterStatus adapterStatus = getAdapterStatus(namespace, adapterName);
+    AdapterSpecification adapterSpec = getAdapter(namespace, adapterName);
     if (adapterStatus != AdapterStatus.STOPPED) {
       throw new CannotBeDeletedException(Id.Adapter.from(namespace, adapterName));
     }
     store.removeAdapter(namespace, adapterName);
 
-    // TODO: Delete the application if this is the last adapter
+    // delete the template if this was the last adapter of that type
+    deleteApp(Id.Application.from(namespace, adapterSpec.getTemplate()));
+  }
+
+  /**
+   * deletes the app if it has no associated adapters
+   * @param templateId the {@link Id.Application} of the template
+   */
+  private void deleteApp(Id.Application templateId) {
```

Comment:
Not sure if this is enough. Take a look at the AppLifecycleHttpHandler and see what all we when we delete an app. Shouldn't we cover all that?